### PR TITLE
Word wrap property for textbox/axis labels

### PR DESCRIPTION
**Pull request by zBritva**
PR for implementing word wrapping features mentioned in: https://github.com/microsoft/charticulator/issues/283, https://github.com/microsoft/charticulator/issues/271

image

Textbox properties
image

Axis properties
image

### CODE REVIEW

File: src/core/prototypes/marks/textbox.ts [hunk 1]
```diff
@@ -277,58 +318,173 @@ export class TextboxElementClass extends EmphasizableMarkClass<
...
+      // if option enabled and no space for rest of text, set limit of lines count
+      if (!props.overFlow) {
+        maxLines = Math.floor(Math.abs(attrs.y2 - attrs.y1) / height);
+      }
+
+      let textContentList = [textContent];
+      // auto wrap text content
+      if (props.wordWrap) {
+        textContentList = splitByWidth(
+          attrs.text.replace(/\n/g, "\\n"),
```
Comment (danmarshall) [inline src/core/prototypes/marks/textbox.ts:380]: Can you make a set of functions for these `.replace` calls

File: src/core/prototypes/plot_segments/axis.ts [hunk 2]
```diff
@@ -362,26 +366,70 @@ export class AxisRenderer {
...
-        g.elements.push(gText);
+        if (style.wordWrap || tick.label.split(/\\n/g).length > 0) {
+          let textContent: string[] = splitByWidth(
+            tick.label.replace(/\n/g, "\\n").replace(/\t/g, "\\t"),
+            maxTickDistance,
+            10000,
+            style.fontFamily,
+            style.fontSize
+          );
+          textContent = textContent.flatMap((line) => line.split(/\\n/g));
```
Comment (danmarshall) [inline src/core/prototypes/plot_segments/axis.ts:377]: Let's also add a function for the `.split`

**Review by danmarshall — CHANGES_REQUESTED** [review]

File: src/core/prototypes/marks/textbox.attrs.ts [hunk 3]
```diff
@@ -72,4 +72,5 @@ export interface TextboxElementProperties extends ObjectProperties {
   paddingY: number;
   alignX: "start" | "middle" | "end";
   alignY: "start" | "middle" | "end";
+  wordWrap: boolean;
```
[the remark this answers is not in the document]
Comment (zBritva) [reply src/core/prototypes/marks/textbox.attrs.ts:75]: [Default value is false](https://github.com/microsoft/charticulator/pull/358/files#diff-b4f2e513a1c89f7210a8d05c5260e84f66df1e625b40d3b11e3881b634fa696eR51).

File: src/core/prototypes/marks/textbox.ts [hunk 4]
```diff
@@ -360,7 +360,7 @@ export class TextboxElementClass extends EmphasizableMarkClass<
         } as Graphics.TextOnPath;
       }
     };
-    const textContent = attrs.text.replace(/\\n/g, "\n");
+    const textContent = replaceNewLineBySymbol(attrs.text);
     if (
       (textContent && textContent.split(/\n/g).length > 1) ||
```
Comment (danmarshall) [inline src/core/prototypes/marks/textbox.ts:365]: should this split call the function?

**Comment by zBritva** [comment]
@danmarshall Could you review the changes? Thanks!

File: src/core/prototypes/plot_segments/axis.ts [hunk 5]
```diff
@@ -485,33 +588,89 @@ export class AxisRenderer {
...
-        0,
-        2
-      );
-      const gt = makeGroup([
-        makeLine(0, 0, 0, style.tickSize * side, lineStyle),
-        makeText(textX, textY, tick.label, style.fontFamily, style.fontSize, {
-          fillColor: style.tickColor,
-        }),
-      ]);
+      const lablel = tick.label && tick.label.replace(/\n/g, "\\n");
```
Comment (danmarshall) [inline src/core/prototypes/plot_segments/axis.ts:601]: function

**Review by danmarshall — CHANGES_REQUESTED** [review]
Just a few more function calls left.

**Review by danmarshall — APPROVED** [review]

**Comment by filsanyousuf** [comment]
How do I use this fix in charticulator?!?

**Comment by zBritva** [comment]
@filsanyousuf you need to build own version of Charticulator to use this feature or wait deployment master branch to site (no ETA when it will be deployed)

**Comment by filsanyousuf** [comment]
> @filsanyousuf you need to build own version of Charticulator to use this feature or wait deployment master branch to site (no ETA when it will be deployed)

Is there a manual on how I do this? Im new to all this..

**Comment by bongshin** [comment]
This feature is deployed in the latest release (Version 2.0.0).

**Pull request merged**